lib/rebaser-server/src/server/core_loop.rs

```diff
@@ -0,0 +1,191 @@
+    stream: Stream,
+    mut shutdown_watch_rx: watch::Receiver<()>,
+) {
+    let mut stream = stream.take_until_if(Box::pin(shutdown_watch_rx.changed().map(|_| true)));
+
+    while let Some(unprocessed_message) = stream.next().await {
+        let message = match unprocessed_message {
+            Ok(processed) => processed,
+            Err(err) => {
+                error!(error = ?err, "error when pull message off stream");
+                continue;
+            }
+        };
+
+        if let Err(err) = message.ack_with(AckKind::Progress).await {
+            error!(error = ?err, "could not ack with progress, going to continue anyway");
+        }
+
+        // Deserialize the message payload so that we can process it.
+        let request_message: RequestRebaseMessage =
+            match serde_json::from_slice(message.message.payload.to_vec().as_slice()) {
+                Ok(deserialized) => deserialized,
+                Err(err) => {
+                    error!(error = ?err, ?message, "failed to deserialize message payload");
+                    continue;
+                }
+            };
+
+        // Pull the reply subject off of the message.
+        let reply_subject = if let Some(headers) = &message.headers {
+            if let Some(value) = headers.get(REPLY_SUBJECT_HEADER_NAME.clone()) {
+                value.to_string()
+            } else {
+                // NOTE(nick): we may actually want to process the message anyway, but things would be super messed up
+                // at that point... because no one should be sending messages exterior to rebaser clients.
+                error!(
+                    ?message,
+                    "no reply subject found in headers, skipping messages because we cannot reply"
+                );
+                continue;
+            }
+        } else {
+            // NOTE(nick): we may actually want to process the message anyway, but things would be super messed up
+            // at that point... because no one should be sending messages exterior to rebaser clients.
+            error!(
+                ?message,
+                "no headers found, skipping message because we cannot reply"
+            );
+            continue;
+        };
+
+        let mut ctx = match ctx_builder.build_default().await {
+            Ok(ctx) => ctx,
+            Err(err) => {
+                // NOTE(nick): we may actually want to bubble up the failure here.
+                error!(error = ?err, "unable to build dal context");
+                continue;
+            }
+        };
+        ctx.update_visibility(Visibility::new_head(false));
+        ctx.update_tenancy(Tenancy::new(WorkspacePk::NONE));
+
+        // TODO(nick): decide if we want this to remain blocking.
+        let start = Instant::now();
+        perform_rebase_and_reply_infallible(&mut ctx, request_message, reply_subject).await;
+        info!("perform rebase and reply total: {:?}", start.elapsed());
```

Comment:
We may want to spawn off tasks. This is where the new code is likely most subject to change.

We need a [work queue retention policy](https://docs.nats.io/using-nats/developer/develop_jetstream/model_deep_dive) and decide how we want to handle "acking" in progress.

For now, we assume we have exactly one rebaser server and that it will block on rebases and perform them in sequence.